File: backend/app/domain/mappers.py
"""Domain entity -> API DTO mappers."""
from app.domain.models import Lesson
from app.domain.schemas import AssetRefOut, LessonManifest, SegmentOut
from app.repositories.base import AssetRepository


def to_manifest(lesson: Lesson, assets: AssetRepository) -> LessonManifest:
    return LessonManifest(
        id=lesson.id,
        topic=lesson.topic,
        title=lesson.title,
        language=lesson.language,
        voice=lesson.voice,
        total_duration_ms=lesson.total_duration_ms,
        created_at=lesson.created_at,
        svg=AssetRefOut(
            asset_id=lesson.svg_asset_id,
            url=assets.url_for(lesson.svg_asset_id),
        ),
        segments=[
            SegmentOut(
                index=seg.index,
                text=seg.text,
                svg_element_ids=seg.svg_element_ids,
                audio=AssetRefOut(
                    asset_id=seg.audio_asset_id,
                    url=assets.url_for(seg.audio_asset_id),
                ),
                duration_ms=seg.duration_ms,
            )
            for seg in lesson.segments
        ],
    )

File: android/app/src/main/java/com/tutorai/app/data/remote/Mappers.kt
package com.tutorai.app.data.remote

import com.tutorai.app.data.remote.dto.LessonManifestDto
import com.tutorai.app.data.remote.dto.SegmentDto
import com.tutorai.app.domain.model.Lesson
import com.tutorai.app.domain.model.Segment

/** Resolve a possibly-relative asset path (e.g. "/api/v1/assets/..") against the base URL. */
internal fun resolveUrl(baseUrl: String, path: String): String =
    if (path.startsWith("http")) path
    else baseUrl.trimEnd('/') + "/" + path.trimStart('/')

fun LessonManifestDto.toLesson(baseUrl: String): Lesson = Lesson(
    id = id,
    topic = topic,
    title = title,
    totalDurationMs = total_duration_ms,
    svgUrl = resolveUrl(baseUrl, svg.url),
    segments = segments.map { it.toSegment(baseUrl) },
)

fun SegmentDto.toSegment(baseUrl: String): Segment = Segment(
    index = index,
    text = text,
    svgElementIds = svg_element_ids,
    audioUrl = resolveUrl(baseUrl, audio.url),
    durationMs = duration_ms,
)
